### DISCUSSION TITLE
gun_pool feedback

### DISCUSSION THREAD

**Discussion by eproxus** [ninenines/gun#310] [General]
While looking at the new gun_pool implementation I had some questions / feedback:
1. A pool is created with `Host` + `Port` but APIs like `info/1` take an argument `Authority` instead. How is one supposed to compute this if you want to e.g. start a pool and then get its info (without having to store the pid somewhere)?
2. Looking at how `Authority` is calculated internally, it seems that the same authority will be created for both `Host` + 80 and `Host` + 443, e.g. `start_pool("example.com", 80)` and `start_pool("example.com", 443)` will both generate a value of `"example.com"`. Is this correct? Does Gun start a pool that can handle both TCP and TLS requests, or will it break if called with the type which it wasn't started with?
3. The pool lookup when performing a request looks for the hardcoded `<<"host">>` header, but an HTTP header can have any case so this fails if the user supplies e.g. `<<"Host">>`. Probably some case conversion has to be done here?
4. There is a weird mismatch between the types of the `Host` argument when creating a pool (which has to be an inet string) and the `<<"host">>` header which can be a binary. As a user it feels odd when they have an implicit connection but they're different types. In my option, a better API would allow binaries for the `Host` argument and internally make it a list if needed by inet.

**Comment by essen** [comment]
1. Right I don't think a function is publicly available for this currently. The relevant code to compute it is in https://github.com/ninenines/gun/blob/master/src/gun_http.erl#L667-L679 and should probably be made available publicly.
2. Yes it's correct. I'm not sure how it will behave in that case. That's a good edge case to look for. If it can't work it can just reject the second pool. If both 80 and 443 are necessary separate scopes can be used anyway. There's more possible edge cases related to tunnels and whatnot.
3. Gun expects header names provided by the user to be all lowercase, like Cowboy. Conversion is only done on header names coming from the network. Lowercase is required in HTTP/2+.
4. Authority and TCP host/port are not the same concept though? Authority (provided in the host header) is an HTTP concept not directly related to the real TCP host/port. In fact some users may want to use a different authority than what they connect to, for example connect to localhost but use example.org as authority. Or use multiple authorities with the same server (if the server has multiple authorities configured). Right now we don't have an option to override the pool used but we could accept the pid when it is known.

**Reply by eproxus** [reply]
I'm not sure I follow. With the current API it would be impossible to send a request to localhost with the `host` header set to `example.org`? (because it would select the pool going to `example.org` or if it doesn't exist, return an error?)

**Reply by essen** [reply]
If you connect to localhost explicitly and want to set the host header to example.org it's not going to find the pool yes.

**Reply by eproxus** [reply]
I find it quite unintuitive that pools are created with a TCP/TLS host + port tuple, but then accessed via the `host` header which is not the same thing at all, as you correctly point out.

**Comment by eproxus** [comment]
Another issue I founds is that it seems `gun_pool` sets its own event handler in the `conn_opts`. If one wants to collect metrics from connections via a custom event handler, that is currently not possible? I.e. the pool implementation will replace any user event handler with its own internal one?

**Reply by essen** [reply]
It doesn't override it just puts itself between Gun and your own event handler.

**Reply by eproxus** [reply]
Ah, you're right of course, thanks! I misread the implementation when looking through the source code. 👍

**Reply by eproxus** [reply]
Follow-up question would be: how do deal with pool events themselves? We'd like to collect metrics from pools, is this possible in some way currently?

**Reply by essen** [reply]
No because I don't know what it would look like currently. The pools themselves don't do much. You can attach some extra info to the event handler state of the connection to identify the pool it belongs to as well. Not sure what else is needed.

**Reply by eproxus** [reply]
Metrics we would be interested in is amount of busy/free connections, connection failures, configured pool sizes etc. So any events that would expose those would be helpful.

If it is possible to detect which pool a connection comes from, a possible solution would be to manually poll pool info on every connection start / stop etc. (which is probably when events would be generated internally in the pool anyway) and report metrics for those. Thoughts?

**Reply by essen** [reply]
Busy/free is based on whether the connections in the pool have available space to open streams so it's a bit complicated to track. Something like a rate of failure to checkout a connection would probably work better.

Total pool sizes right now are fixed IIRC but it'll make sense to have when pools grow or shrink in size. In the meantime I believe the info function gives the value. On top of that you could count the number of connection success/failure a pool has, connection failures and such are at the connection-level.

**Reply by eproxus** [reply]
Does it make sense to call `gun_pool:info/1` on each request to produce statistics about the pool? Or is it too heavyweight?

**Reply by essen** [reply]
It depends on how many requests you do I guess. It is not an expensive call on its own.

**Reply by eproxus** [reply]
Which error is returned when all streams are busy in the pool? When trying to use it we sometimes get `{error, {stream_error, closed}}`, does this mean the pool is busy or is it something else?

**Reply by essen** [reply]
I guess you got a connection that ended up closing before you could do something with it, or while doing something. It happens. You may get other errors in that scenario. What should be done in that case is try again. The pool cannot guarantee that connections stay up.

**Reply by eproxus** [reply]
That sounds likely, yes. How can we detect if the pool itself is out of free connections/streams?

**Reply by essen** [reply]
If you mean when doing requests you should hit this path: https://github.com/ninenines/gun/blob/master/src/gun_pool.erl#L367-L368
